juju_spell/commands/grant.py

```
@@ -14,21 +14,83 @@
 """Command to grant permission for user."""
-from typing import Any
+import logging
+from typing import Any, List, Optional
 
 from juju.controller import Controller
 
 from juju_spell.commands.base import BaseJujuCommand
+from juju_spell.exceptions import JujuSpellError
 
 __all__ = ["GrantCommand", "ACL_CHOICES"]
 
-ACL_CHOICES = ["login", "add-model", "superuser"]
+CONTROLLER_ACL_CHOICES = ["login", "add-model", "superuser"]
+MODEL_ACL_CHOICES = ["read", "write", "admin"]
+
+ACL_CHOICES = CONTROLLER_ACL_CHOICES + MODEL_ACL_CHOICES
+
+logger = logging.getLogger(__name__)
 
 
 class GrantCommand(BaseJujuCommand):
     """Grant permission for user."""
 
-    async def execute(self, controller: Controller, *args: Any, **kwargs: Any) -> bool:
+    async def execute(
+        self,
+        controller: Controller,
+        *args: Any,
+        models: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> bool:
         """Execute."""
-        result: bool = await controller.grant(username=kwargs["user"], acl=kwargs["acl"])
-        return result
+        acl = kwargs["acl"]
+        controller_acl: str
+        model_acl: str
+        if acl in CONTROLLER_ACL_CHOICES:
+            controller_acl = acl
+        else:
+            controller_acl = "login"
+
+        if acl in MODEL_ACL_CHOICES:
+            model_acl = acl
+        elif acl == "superuser":
+            model_acl = "admin"
+        else:
+            model_acl = "read"
+
+        logger.info(
+            "Start grant permission %s on controller %s for user %s",
+            controller_acl,
+            controller.uuid,
+            kwargs["user"],
+        )
+        result: bool = await controller.grant(username=kwargs["user"], acl=controller_acl)
+        if not result:
+            raise JujuSpellError(
+                f"Grant user {kwargs['user']} permission {controller_acl}"
+                f" on controller {controller.uuid} fail."
+            )
```

I'm not sure if we should failed here or not. Maybe here yes and if grant to model failed not? What do you think @jneo8 and @esunar ? From user perspective.

---

Updated.

After I checked the python-libjuju. It will raise error from the lib directly. So I remove the raise part here.

---

Is that behavior which we want? If grant permission for single model failed (model is broken or I do not know) we will end up with user, which have permission only for some models

---

In the newest commit. We provide a argument `--overwrite`. It will ignore the exception and continue to next step, otherwise will get exception.